### PR TITLE
Improve annotations in automl and ml modules

### PR DESCRIPTION
**Pull request by coffepowered**
## Why are these changes needed?

Hi, this should help new FLAML's contributors familiarise themselves with the code.

Annotation is still not comprehensive but helps (me, at least) understanding:

- when variables named `estimator` are actual instances of some BaseEstimator subclass 
- what's the "state" being passed in "_compute_with_config_base" (it's AutoMLState to my understanding, is there any case in which it could be something different?). Notice that in order to annotate this had to import annotations from "future"
-  when "metric" and other variable names occurring frequently are strings without relying on the variable name itself (e.g. metric_str).

## Related issue number
The import from `future` has been briefly discussed in #907 .
I am sorry the modifications are in 2 modules, but had to look at both to better understand the expected types.

## Checks

- [x] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [x] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: flaml/automl/ml.py [hunk 1]
```diff
@@ -424,14 +433,14 @@ def _eval_estimator(
 
 def get_val_loss(
     config,
-    estimator,
+    estimator: Optional[EstimatorSubclass],
```
Comment (sonichi) [inline flaml/automl/ml.py:436]: Not optional.

File: flaml/automl/ml.py [hunk 2]
```diff
@@ -608,17 +617,17 @@ def compute_estimator(
...
-    n_jobs=1,
-    estimator_class=None,
-    cv_score_agg_func=None,
-    log_training_metric=False,
-    fit_kwargs={},
+    n_jobs: int = 1,
+    estimator_class: Optional[EstimatorSubclass] = None,
+    cv_score_agg_func: Optional[callable] = None,
+    log_training_metric: Optional[bool] = False,
+    fit_kwargs: dict = {},
```
Comment (sonichi) [inline flaml/automl/ml.py:630]: Optional

File: flaml/automl/ml.py [hunk 3]
```diff
@@ -608,17 +617,17 @@ def compute_estimator(
...
+    estimator_name: str,
+    eval_method: str,
+    eval_metric: Union[str, Callable],
     best_val_loss=np.Inf,
-    n_jobs=1,
-    estimator_class=None,
-    cv_score_agg_func=None,
-    log_training_metric=False,
-    fit_kwargs={},
+    n_jobs: int = 1,
```
Comment (sonichi) [inline flaml/automl/ml.py:626]: Optional

File: flaml/automl/ml.py [hunk 4]
```diff
@@ -678,18 +687,18 @@ def compute_estimator(
...
+    config_dic: dict,
     X_train,
     y_train,
-    task,
-    estimator_name,
-    n_jobs=1,
-    estimator_class=None,
+    task: str,
+    estimator_name: str,
+    n_jobs: int = 1,
```
Comment (sonichi) [inline flaml/automl/ml.py:695]: Optional

File: flaml/automl/ml.py [hunk 5]
```diff
@@ -678,18 +687,18 @@ def compute_estimator(
...
+    estimator_class: Optional[EstimatorSubclass] = None,
     budget=None,
-    fit_kwargs={},
+    fit_kwargs: dict = {},
```
Comment (sonichi) [inline flaml/automl/ml.py:698]: Optional

Comment (coffepowered) [reply flaml/automl/ml.py:698]: Hey, thanks for all your feedbacks!
This was slightly tricky to handle. Take a look at my proposed solution in the updated version

File: flaml/automl/ml.py [hunk 6]
```diff
@@ -678,25 +692,28 @@ def compute_estimator(
...
+    config_dic: dict,
     X_train,
     y_train,
-    task,
-    estimator_name,
-    n_jobs=1,
-    estimator_class=None,
+    task: str,
+    estimator_name: str,
+    n_jobs: Optional[int] = 1,
```
Comment (sonichi) [inline flaml/automl/ml.py:700]: Does the comment in 628 apply to here as well?

**Review by sonichi — COMMENTED** [review]
A test fails. Please check whether it's caused by a change in this PR.

File: flaml/automl/ml.py (hunk 6 above)
Comment (coffepowered) [reply flaml/automl/ml.py:700]: It does. It's just a reminder reported from class `SKLearnEstimator`. Not sure what should I do: should I remove the inline comment altogether or just copy it once again here?

**Comment by ZviBaratz** [comment]
@coffepowered well done moving forward with this. I'm sorry to say that I have a pretty busy couple of weeks ahead of me, so I don't think I will be able to give a proper review soon.

File: flaml/automl/ml.py (hunk 6 above)
Comment (sonichi) [reply flaml/automl/ml.py:700]: > It does. It's just a reminder reported from class `SKLearnEstimator`. Not sure what should I do: should I remove the inline comment altogether or just copy it once again here?

You can copy it again here.

Comment (sonichi) [reply flaml/automl/ml.py:700]: > Also, there's something weird going on - I am not able to reproduce the pipeline error locally. Seems that PR #918 fails in the same test, do you have any idea why?

@int-chaos is investigating it. Most likely some dependencies changed versions.

Comment (sonichi) [reply flaml/automl/ml.py:700]: > > It does. It's just a reminder reported from class `SKLearnEstimator`. Not sure what should I do: should I remove the inline comment altogether or just copy it once again here?
> 
> You can copy it again here.

This has not been addressed.

File: flaml/automl/automl.py [hunk 7]
```diff
@@ -344,7 +345,9 @@ def _prepare_sample_train_data(self, sample_size):
         return sampled_X_train, sampled_y_train, sampled_weight, groups
 
     @staticmethod
-    def _compute_with_config_base(config_w_resource, state, estimator, is_report=True):
+    def _compute_with_config_base(
+        config_w_resource: dict, state: AutoMLState, estimator, is_report=True
```
Comment (qingyun-wu) [inline flaml/automl/automl.py:349]:
```suggestion
        config_w_resource: dict, state: AutoMLState, estimator: str, is_report: bool = True
```

**Review by sonichi — APPROVED** [review]

**Review by qingyun-wu — APPROVED** [review]

**Pull request merged**